test/e2e/run/e2e.go

```diff
@@ -113,34 +113,35 @@ func (e *E2E) Run(ctx context.Context) (E2EResult, error) {
 		return E2EResult{Phases: phases}, err
 	}
 
-	// After this point we are going to run as much as we can regardeless of potential errors
-	// and will collect all errors to return them at the end
-	allErrors := []error{}
-
-	// save 2 minutes for reporting/s3 uploading
+	// save 5 minutes for cleanup and 2 minutes for reporting/s3 uploading
 	runTestsDeadline := deadline.Add(-reportingBuffer)
+	if !e.SkipCleanup {
+		runTestsDeadline = runTestsDeadline.Add(-cleanupBuffer)
+	}
 	runTestsCtx, runTestsCancelFunc := context.WithDeadline(ctx, runTestsDeadline)
 	defer runTestsCancelFunc()
-	runTestsPhases, err := e.runTests(runTestsCtx)
-	if err != nil {
-		allErrors = append(allErrors, err)
-	}
+	runTestsPhases := e.runTests(runTestsCtx)
 	phases = append(phases, runTestsPhases...)
 
-	e2eResult, err := e.parseReport(ctx)
-	phases, err = phaseCompleted(phases, phaseNameParseReport, "parsing report", err)
-	if err != nil {
-		allErrors = append(allErrors, err)
-	}
+	// save 2 minutes for reporting/s3 uploading
+	cleanupDeadline := deadline.Add(-reportingBuffer)
+	cleanupCtx, cleanupCancelFunc := context.WithDeadline(ctx, cleanupDeadline)
+	defer cleanupCancelFunc()
+	cleanupPhases := e.runCleanup(cleanupCtx)
+	phases = append(phases, cleanupPhases...)
 
-	err = e.uploadArtifactsToS3(ctx)
-	phases, err = phaseCompleted(phases, phaseNameUploadArtifactsToS3, "uploading logs to s3", err)
-	if err != nil {
-		allErrors = append(allErrors, err)
-	}
+	e2eResult, parsePhases := e.parseReport(ctx)
+	phases = append(phases, parsePhases...)
+
+	uploadPhases := e.uploadArtifactsToS3(ctx)
+	phases = append(phases, uploadPhases...)
 
 	e2eResult.Phases = phases
-	return e2eResult, errors.Join(allErrors...)
+	var allErrors error
+	for _, phase := range phases {
+		allErrors = errors.Join(allErrors, phase.Error)
```

Comment:
should we check if err != nil? i'm assuming that errors Join only adds non nil errors but it seems more explicit to maybe do it here?